### ISSUE TITLE
Support individual files?

### ISSUE THREAD

**Issue by marksteward** [mxmlnkn/ratarmount#31]
It'd be really nice to be able to mount single files in a directory so that file is exposed decompressed without having to be a tar.

**Comment by mxmlnkn** [comment]
Interesting use-case. So you want to mount a file.bz2 or file.gz. That could prove handy when adding some automated tests when parallelizing the bz2 decoder.

Suppose you have `./foo.bz2`, would it work for you if it then would be accessible in `./foo/foo` instead of directly `./foo`? I think that would be much easier to integrate than the latter. Thinks like `ratarmount file1.bz2 file2.bz2 file3.bz2 file4.bz2 mountpoint` should also work rather easily.

I'm just thinking out loud here, but issues to look into would be:

 - Can the FUSE mount point itself be a file instead of a folder? If this works, e.g., by simply returning the correct attributes, then it still would pose problems as to how to integrate it into the existing TarMount class. I fear it might be easier to basically do this from scratch in a different logic path. And here I just removed the old index serialization backend in version 0.5.0 to reduce the code complexity a bit. But if I don't integrate it, it might not work together with e.g. union mounting and such.
 - How should it behave for .tar.bz2 files. Should there be a command line option to only decompress the bz2 layer instead of tar and bz2
 - How should this combine with the union mounting interface. The above single flag definitely wouldn't work if you wanted to union mount a tar.bz2, for which the tar should also be decompressed, and another .tar.bz2, for which only the bz2 layer should be decompressed. The current CLI would make per-file flags a bit harder to do. But, it shouldn't be impossible and others like ImageMagick and ffmpeg have similarly complex CLIs.
 - Almost a non question, but what should the file name be because bz2 stores no original file name information? The only solution is to use the the bz2 filename without the extension assuming it has one. If not, then the mount point might clash and has to be specified manually.
 - How might the saved single bz2 index collide with the preexisting TAR-centric index SQLite tables?

**Comment by marksteward** [comment]
> Suppose you have ./foo.bz2, would it work for you if it then would be accessible in ./foo/foo instead of directly ./foo? I think that would be much easier to integrate than the latter. Thinks like ratarmount file1.bz2 file2.bz2 file3.bz2 file4.bz2 mountpoint should also work rather easily.

Yes, that makes sense to me. I don't think FUSE supports mounting files.

> How should it behave for .tar.bz2 files. Should there be a command line option to only decompress the bz2 layer instead of tar and bz2

Hmm, I hadn't considered this. You could change the recursive mode to be "untar depth", defaulting to 1. This feature would then become the special case of 0.

> How should this combine with the union mounting interface.

I'm a bit wary of following ImageMagick's example ;) Maybe complex situations could be worked around with symlinks?

> How might the saved single bz2 index collide with the preexisting TAR-centric index SQLite tables?

I figured it would just skip the index.

My use case was loading up a sqlite3 DB that's distributed as a .bz2. What I haven't checked is whether the python Bzip library allows efficient random access within the compressed stream, or whether it ends up decompressing it all to memory. Do you happen to have covered this in your investigations?

**Comment by mxmlnkn** [comment]
> My use case was loading up a sqlite3 DB that's distributed as a .bz2. What I haven't checked is whether the python Bzip library allows efficient random access within the compressed stream, or whether it ends up decompressing it all to memory. Do you happen to have covered this in your investigations?

As far as I know Python's bzip2 module has no seeking support for bzip2, no seeking back and no seeking ahead. It also has no interface to query and possibly set or save the bzip2 block boundaries. I really tried long to abuse Python's bzip2 module even with possibly internal interfaces. However, I had to use a custom modified bzip2 decoder in the end. I think it doesn't even decompress everything to memory. If you seek to somewhere, it decompress all blocks before it block-wise and block-wise discards the results.

**Comment by marksteward** [comment]
> However, I had to use a custom modified bzip2 decoder in the end. I think it doesn't even decompress everything to memory. If you seek to somewhere, it decompress all blocks before it block-wise and block-wise discards the results.

Ahh nice!

**Comment by mxmlnkn** [comment]
I hacked something together in 27c8ca396adc3fdb53849349af127e3a998d611a. You can try it out using `pip install git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount`. I didn't add tests yet, though and the last polish might be missing. It worked pretty good in my manual tests, though.

**Comment by marksteward** [comment]
Heh, it's "discovered" 34 invalid filenames inside the bzip2. I patched line 412:

```
412  	          for tarInfo in []: #loadedTarFile:
```

and it worked fine. Nice work!

**Comment by mxmlnkn** [comment]
That is unexepected. That would mean that the tarfile module finds some chunks of data which it thinks are valid TAR data in your uncompressed data, which shouldn't be. I think because of the `ignore_zeros = True` option tarfile might be a bit [too fault-tolerant](https://bugs.python.org/issue40757) not just skipping zeros but basically any chunk of invalid data. Instead of trying to open it as a TAR and checking for errors, I'll have try to forward the results of the file type analysis, which does not have the `ignore_zeros` option turned to true, deeper (from `_detectCompression` to `createIndex`). I could reproduce a similar problem with a 10GB bz2 compressed sqlite but not with a smaller 100MB sqlite file. I also noticed that the progress info output shouldn't even appear if there are no files in a TAR or no TAR at all because the progress is only triggered when a file is detected.

I pushed a version which should behave much more benign and should fix both issues.

Btw during testing, I think I stumbled upon [this issue](https://stackoverflow.com/questions/41396797/cant-query-sqlite-file-located-in-read-only-directory). Looks like some of my SQLite test databases were in WAL mode and so I couldn't get them to open in ratarmount no matter what. This might be a good reason to implement a writable overlay layer even if those written changes wouldn't be written back to the compressed TAR. However, the SQLite databases created by ratarmount were perfectly readable when bz2 compressed and then mounted with ratarmount itself and I have a 3GB ratarmount index, I hope that should prove complex enough a sample.
